x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/workspace_panel.test.tsx

```diff
@@ -233,39 +237,90 @@ describe('workspace_panel', () => {
-                "type": "expression",
-              },
-              Object {
-                "chain": Array [
-                  Object {
-                    "arguments": Object {},
-                    "function": "datasource2",
-                    "type": "function",
-                  },
-                ],
-                "type": "expression",
-              },
-            ]
-        `);
+                  Array [
+                    Object {
+                      "chain": Array [
+                        Object {
+                          "arguments": Object {},
+                          "function": "datasource",
+                          "type": "function",
+                        },
+                      ],
+                      "type": "expression",
+                    },
+                    Object {
+                      "chain": Array [
+                        Object {
+                          "arguments": Object {},
+                          "function": "datasource2",
+                          "type": "function",
+                        },
+                      ],
+                      "type": "expression",
+                    },
+                    Object {
+                      "chain": Array [
+                        Object {
+                          "arguments": Object {},
+                          "function": "datasource2",
+                          "type": "function",
+                        },
+                      ],
+                      "type": "expression",
+                    },
+                  ]
+            `);
+  });
+
+  it('should run the expression again if the date range changes', async () => {
+    const framePublicAPI = createMockFramePublicAPI();
+    framePublicAPI.datasourceLayers = {
+      first: mockDatasource.publicAPIMock,
+    };
+    mockDatasource.getLayers.mockReturnValue(['first']);
+
+    mockDatasource.toExpression
+      .mockReturnValueOnce('datasource')
+      .mockReturnValueOnce('datasource second');
+
+    expressionRendererMock = jest.fn(_arg => <span />);
+
+    instance = mount(
+      <InnerWorkspacePanel
+        activeDatasourceId={'mock'}
+        datasourceStates={{
+          mock: {
+            state: {},
+            isLoading: false,
+          },
+        }}
+        datasourceMap={{
+          mock: mockDatasource,
+        }}
+        framePublicAPI={framePublicAPI}
+        activeVisualizationId="vis"
+        visualizationMap={{
+          vis: { ...mockVisualization, toExpression: () => 'vis' },
+        }}
+        visualizationState={{}}
+        dispatch={() => {}}
+        ExpressionRenderer={expressionRendererMock}
+      />
+    );
+
+    // "wait" for the expression to execute
+    await waitForPromises();
+    instance.update();
+
+    expect(expressionRendererMock).toHaveBeenCalledTimes(1);
+
+    framePublicAPI.dateRange = { fromDate: 'now-90d', toDate: 'now-30d' };
+    instance.setProps({ visualizationState: {} });
```

Comment:
nit: The way the update is triggered is slightly confusing (the visualization state didn't actually change, yet it's triggering the update). Why not something like `instance.setProps({ framePublicAPI: {...framePublicAPI, dateRange: { fromDate: 'now-90d', toDate: 'now-30d' } } });`?

Reply:
Good point